PyRoute/TradeCodes.py

```diff
@@ -80,6 +123,103 @@ def __init__(self, initial_codes):
         self.dcode = list(self.dcode) + self.colony + self.owner
         self.dcode = sorted(self.dcode)
 
+    def _preprocess_initial_codes(self, initial_codes):
+        raw_codes = initial_codes.split()
+        # look for successive codes that should be part of the same name (eg  "(Carte" , then "Blanche)", becoming
+        # "(Carte Blanche)", and bolt them together
+        num_codes = len(raw_codes)
+        codes = []
+        for i in range(0, num_codes):
+            raw = raw_codes[i]
+            if '' == raw:
+                continue
+            if ')' == raw:
+                continue
+            if raw.startswith('Di('):
+                codes.append(raw)
+                continue
+            if 7 < len(raw) and '(' == raw[0] and ')' == raw[-2]:  # Let older-style sophont codes through
+                codes.append(raw)
+                continue
+            if not raw.startswith('(') and '(' in raw and raw.endswith(')'):
+                continue
+            if not raw.endswith(')') and ')' in raw and raw.startswith('(') and 7 > len(raw):
+                continue
+            if not raw.startswith('(') and not raw.endswith(')') and '(' in raw and ')' in raw:
+                continue
+            if 7 == len(raw) and '(' == raw[0] and ')' == raw[5]:  # Let preprocessed sophont codes through
+                codes.append(raw)
+                continue
+            if not raw.startswith('('):
+                codes.append(raw)
+                continue
+            if raw.endswith(')'):
+                codes.append(raw)
+                continue
+            if i < num_codes - 1:  # Try to join cases like '(Carte' and 'Blanche)'
+                next = raw_codes[i + 1]
+                if next.endswith(')'):
+                    combo = raw + ' ' + next
+                    codes.append(combo)
+                    raw_codes[i + 1] = ''
+
+        initial_codes = ' '.join(codes)
+
+        return codes, initial_codes
+
+    def _process_homeworld(self, homeworld, homeworlds_found, initial_codes):
+        full_name = re.sub(r'\(([^)]+)\)\d?', r'\1', homeworld)
+        homeworlds_found.append(homeworld)
+        match = re.match(r'\w{,2}\(([^)]{,4})[^)]*\)(\d|W|\?)?', homeworld)
+        if match is None:
+            self.logger.error("Unable to process %s", initial_codes)
+            sys.exit(1)
+        if full_name.startswith("Di"):
+            sophont = "{code: <4}{pop}".format(code=match.group(1), pop='X')
+        else:
+            pop = match.group(2) if match.group(2) else 'W'
+            if homeworld + '?' in initial_codes:
+                homeworld += "?"
+                pop = '0'
+            elif homeworld + str(pop) in initial_codes:
+                homeworld += str(pop)
+            sophont = "({code: <4}){pop}".format(code=match.group(1), pop=pop)
+        sophont = sophont.replace("'", "X")
+        sophont = sophont.replace("!", "X")
+        sophont = sophont.replace(" ", "X")
```

Comment:
I'm trying to remember if the Traveller Wiki does this any more. The reason for this substitution was the wiki used this substitution as a recommendation from another source to create the four character sophont codes. The classic example was the K'kree. Their four character code is "K'kr" but for the wiki it became "KXkr". 

In any case, the root use of this was to allow PyRoute to accept the data from Traveller Map. Then generate output for the Traveller Wiki. This could, without having to maintain in internal list of the short code -> full name mapping, use a template and make the wiki do the lookup in the page. We should be able to remove these three lines and simply let the wiki manage the short codes.

Reply:
As this was triplicated (see your later comments), I've pulled this into its own little method, namely:

```
    def _process_sophont_homeworld(self, code, pop, direct_homeworld=False):
        if direct_homeworld:
            sophont = "({code: <4}){pop}".format(code=code, pop=pop)
        else:
            sophont = "{code: <4}{pop}".format(code=code, pop=pop)
        sophont = sophont.replace("'", "X")
        sophont = sophont.replace("!", "X")
        sophont = sophont.replace(" ", "X")
        self.sophont_list.append(sophont)
        self.homeworld_list.append(sophont)

        return sophont
```